### PR TITLE
Add configuration for prettier code formatting

### PR DESCRIPTION
**Pull request by mramato**
We have always worked incredibly hard to keep the CesiumJS code base clean and well-formatted. We've always believed that a good code base should look like it was written by a single person, even with the ~200 contributors we've had on CesiumJS, we've largely hit and maintained that goal.  But it's been tough.

[prettier](prettier.io) took the JS world by storm a while ago and the time has come for CesiumJS to embrace it and automatically format all (actually most) of our JavaScript code.

At a high level, this PR:

1. Adds prettier and configuration files, but does not format code yet.

2. Configures prettier for Source/Apps/Specs but not anything else. JS code only. If we want to do Markdown and HTML code, we can.  HTML may be a little tough because we have to trim whitespace so we can show it in Sandcastle so it may be better to do that in a separate PR, but I'm open to  opinions.

3. Fix a bad closing div in `CZML Custom Properties.html`  that prettier found while I was experimenting.

4. Adds a pre-commit hook via husky and pretty-quick which will automatically format changed files when they are committed on the client.  This means that you don't even need to THINK about code formatting, it just magically happens when you commit and it's fast.

5. Run `prettier-check` during CI and fail the build if code is not formatted properly.

6. Install eslint config for prettier. This does not enable prettier checking in eslint (it's unusably slow) but instead just makes sure that no eslint rules conflict with prettier formatting.

7. Update the coding guide to remove our manual rules and just mention prettier.

If you want to see the results yourself, just check out this branch, run `npm install` and then run `npm run prettier`.  It will take a minute since it is doing the entire code-base right now but that will allow you to see what the "new" code will look like when we do the initial mass-format.

prettier has a limited number of options specified in `.prettierrc`. I mostly kept he defaults, but when with the following because it's how Cesium has always worked. (Okay the arrowParens once doesn't apply yet, but it will really soon).

```json
{
  "singleQuote": true,
  "tabWidth": 4,
  "trailingComma": "none",
  "arrowParens": "avoid"
}
```

At prettier's suggestion, I kept the default line width to 80, but that's open for debate as well.

While there is no immediate rush to converge on this, I would like to get to an answer sooner rather than later and avoid too much bike shedding.  I honestly don't care about the formatting that much as long as it is consistent.  It will also be easier to run prettier everywhere and then put back any custom formatting people want via [prettier-ignore](https://prettier.io/docs/en/ignore.html)

Once we all agree, we'll do a single commit with the full format in order to make it easy to merge into other branches.  @kring can provide some guidance here since he's done it for other projects.

EDIT: We've since gone back to all of prettier's defaults, as per the discussion below.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @mramato!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/lets-argue-over-minutiae/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by mramato** [comment]
FYI, this branch will fail because `prettier-check` is failing because I haven't actually formatted the code yet.  Depending on the order to do things, I can undo this check and re-add it after we run the format, I just wanted to make sure it does indeed fail.

**Comment by kring** [comment]
Looks good to me. Only thing I'd suggest is we stick to the default settings. Everyone will have to get used to a new style anyway, so no reason to stick to Cesium's past conventions for historical reasons. The default tab width of 2 is actually pretty important because of prettier's hard limit of 80 columns (and I don't recommend increasing this because then it tends to unwrap things you really don't want unwrapped). The defaults are well considered, we shouldn't change them without a very good reason.

File: Documentation/Contributors/CodingGuide/README.md [hunk 1]
```diff
@@ -90,52 +90,9 @@ A few more naming conventions are introduced below along with their design patte
...
-
-    this._ready = false;
-    this._readyPromise = when.defer();
-    // ...
-};
-```
-* In JavaScript code, use single quotes, `'`, instead of double quotes, `"`.  In HTML, use double quotes.
+We use [prettier](https://prettier.io/) to automatically re-format all JS code at commit time, so all of the work is done for you.
 
+* For HTML code, simple keep the existing style. Use double quotes.
```
Comment (OmarShehata) [inline Documentation/Contributors/CodingGuide/README.md:95]: `simple keep` -> `simply keep` ?

Comment (mramato) [reply Documentation/Contributors/CodingGuide/README.md:95]: Yes, thanks.

**Comment by mramato** [comment]
> Only thing I'd suggest is we stick to the default settings.

I'll admit, it pains me a little to agree to this because I hate trailing commas and quotes in JS will take a lot of getting used to, but I'm willing to do it if everyone else is on board.  Ultimately I care about consistency.

> The defaults are well considered, we shouldn't change them without a very good reason.

For anyone that ones more info on this, there is an entire page that goes over why the defaults are what they are and how they work, here: https://prettier.io/docs/en/rationale.html

Who else has feedback? Don't let a good branch name go to waste :smile:

**Comment by GatorScott** [comment]
80 columns and 2-space indent doesn't seem like progress.  Were other alternatives considered?

**Comment by kring** [comment]
> I'll admit, it pains me a little to agree to this because I hate trailing commas and quotes in JS will take a lot of getting used to, but I'm willing to do it if everyone else is on board. Ultimately I care about consistency.

Wow I didn't realize the latest version of Prettier had enabled trailing commas by default 🤯.
So yeah that doesn't sound amazing to me, either. But, as usual, that's mostly cause I'm not used to it, and there are decent reasons to do it, and pretty soon I'll get used to it and it'll seem normal. ;)

Switching Cesium's quotes is kinda nice for folks that switch between JS and C++ (and other languages), though.

> 80 columns and 2-space indent doesn't seem like progress.

Yeah that's what I thought too. This was my number one complaint about prettier before switching to it. I don't mind anymore. Like it, even. There's a good chance it'll happen to you too. :)

> Were other alternatives considered?

Well there are a billion alternatives, many of which would be more similar to our current style. But prettier is taking over the javascript world (and other languages as well), and its well-considered choices are only unfamiliar, never objectively bad. Once you get past the "this looks ugly" feeling, the style starts to feel familiar, and you get used to literally never having to think about formatting _at all_, it's really hard to go back.

**Comment by hpinkos** [comment]
This all looks fine to me

**Comment by mramato** [comment]
> Wow I didn't realize the latest version of Prettier had enabled trailing commas by default .

It's technically "ES5" instead of always: https://prettier.io/docs/en/options.html#trailing-commas (It's also only when multi-line.  So not the worst idea but will definitely take getting used to. They don't technically provide a justification for it on the rationale page either (though the typical one is that it makes copy/paste/adding new values easier.

**Comment by mramato** [comment]
@kring I removed the .prettierrc and went back to all defaults. Oddly it seems to be defaulting to a indent width of 4, not sure why.  Any ideas?

**Comment by hpinkos** [comment]
It looks like prettier is trying to run on the generated shaders and workers?  Do we want to skip those?

And can we pretty please keep single quotes?

**Comment by mramato** [comment]
> It looks like prettier is trying to run on the generated shaders and workers? Do we want to skip those?

It doesn't actually matter, probably better to ignore,

> And can we pretty please keep single quotes?

See the above conversation. We plan on using prettier defaults, which is double quotes.  Yes it's a change that will take getting used to, but it does make sense in that apostrophes are more common than quotations, so in the grand scheme you save a few bytes.

Even if you don't by the justification, what matters in the grand scheme is consistency so there's no good reason to change the default.

**Comment by mramato** [comment]
I wrote up a prettier issue for the tabWidth problem I'm seeing: https://github.com/prettier/prettier/issues/8043

**Comment by hpinkos** [comment]
> Even if you don't by the justification, what matters in the grand scheme is consistency so there's no good reason to change the default.

I thought we at least wanted to override the trailing comma default.  Don't trailing commas kill old versions of IE?  I thought I'm not sure how much we care, but I've had to fix that in order to get the NTS app to build in the past.  (obviously that build is way overdue for an update too).  Also, does adding extra commas add significantly to the size of the final build?

I get that my preference for single quote is largely aesthetic.  I guess I can still type a single quote all the time and prettier will just fix it for me on commit.

**Comment by mramato** [comment]
> Don't trailing commas kill old versions of IE

IE11 is the only version we care about at all these days. I think the trailing commas setting will work with it because it is "ES5" by default, which means not always just when ES5 supports it. Validating that IE11 works is certainly something we should do before we merge this.

**Comment by mramato** [comment]
Okay, made a bunch of improvements so this is now working on the entire codebase instead of just JS code.

1. Deleted editor config, since we now use prettier, it was messing with prettier configuration, which respects it.  See https://github.com/prettier/prettier/issues/8043
2. Format css, html, and md in addition to JS.
3. Remove extra whitespace from Sandcastle examples.
4. Flesh out .prettierignore so that it no longer does generated sharders/workers/Cesium.js

**Comment by mramato** [comment]
There is also now a branch you can look at if you want to see what the formatted code is like: https://github.com/CesiumGS/cesium/tree/test-prettier

**Comment by hpinkos** [comment]
Hate the double quotes.  Hate the 2 space indent.  Hate the 80 character column width.  Hate the trailing comma.  But at least it's all the same and I don't have to think about it? :woman_shrugging:

**Comment by mramato** [comment]
> Hate the double quotes. Hate the 2 space indent. Hate the 80 character column width. Hate the trailing comma. But at least it's all the same and I don't have to think about it?

Pretty much, yes. It just takes time to get used to new formatting. There was a time in my life where I was firmly on the "curly braces are always on a new line" side of the formatting holy wars and I would have told you curly braces on the end/beginning of lines were trash.  Now I feel the complete opposite after working with JavaScript for so long.

So I get where you're coming from, but the JS world has embraced prettier so we should too.  I've already started to get used to the output to be honest.

**Comment by GatorScott** [comment]
Is it easy to disable?  I'm worried about headaches if I'm working on a branch with active CI.  I write some code the way I like, submit it and leave for the day.  I arrive the next morning to CI problems, ok, but when I look at the code to fix it I don't recognize it, slowing me down.

Hey, who moved my cheese?  Too concerned?

**Comment by IanLilleyT** [comment]
Stuff like this worries me: https://github.com/CesiumGS/cesium/blob/d0c8868b990d4efd42bea4da5b3cb9997a640006/Source/Scene/Model.js#L4356-L4369
https://github.com/CesiumGS/cesium/blob/d0c8868b990d4efd42bea4da5b3cb9997a640006/Source/Scene/Model.js#L3896-L3913

Just feels weird in a lot of places. Might force us to change the way we code, like creating more temporary variables to please the formatter.

Our verbose API naming doesn't play too nicely with 80 column limit, especially math stuff.

**Comment by mramato** [comment]
@GatorScott You can set an environment variable `HUSKY_SKIP_HOOKS=1` if you want to disable the git hooks that CesiumJS is using to enforce formatting.  You can also just pass `--no-verify` to `git commit` for a one-off ignore of the commit command.

**Comment by mramato** [comment]
> Just feels weird in a lot of places. Might force us to change the way we code, like creating more temporary variables to please the formatter.

@IanLilleyT I don't disagree with you but I also think that's part of the point. If you have a line that's too long and hard to read, it should probably be broken up anyway.

That being said, you do bring up a good point that Cesium's API is a bit verbose for readability purposes (for example, `boundingSphere` instead of `bs`, `maximum` instead of `max`, etc...) and that lends itself to odd wrapping in places.  The use of statics instead of prototypes also makes for much longer code lines.

I'm personally not against setting the `printWidth` to 100 or 120. because I think the above make for a good enough reason. I would leave all of the other defaults, since there is not as good as a reason for changing them. But there is a certain appeal to just keeping it at 80 and seeing if we adjust to it (not like it would be hard to extend in a month).

Also, keep in mind that we can selectively disable prettier for some parts of the code, such as custom Matrix4 assignment formatting.  This should be exceptionally rare, however, so let's not assume we're going to start disabling prettier everywhere just because we have differing opinions with it.

80 character line width seems to be the biggest sticking point, so if that's where we have to compromise to all agree on using prettier, I'm not going to fight it. Like I said, my biggest goal is to just make the code uniform by default and reduce developer friction.

@kring what are your thoughts here? Do you want to make a case for 80 given Cesium's verbosity and the type of code that @IanLilleyT pointed to?

**Comment by mramato** [comment]
For discussion, I put together another branch that uses a 120 column limit https://github.com/CesiumGS/cesium/tree/test-prettier-120 which avoids all of the odd breaking that was pointed out.

**Comment by kring** [comment]
@IanLilleyT:

> Stuff like this worries me:

Believe it or not, I find that "weird" code way easier to read than the original. :) Once you get used to how prettier indents things, the indentation tells you the structure / precedence of the code, 100% reliably every time.

@mramato: 

> Deleted editor config

You might want to leave it but change the indent to 2. That way VS Code will know that new files should be indented with two spaces.

> I'm personally not against setting the printWidth to 100 or 120. because I think the above make for a good enough reason.

I thought so too at first, but after I tried it I changed my mind. Prettier ends up unwrapping stuff and most lines end up too long. The prettier docs nail it IMO:
https://prettier.io/docs/en/options.html#print-width

> Also, keep in mind that we can selectively disable prettier for some parts of the code, such as custom Matrix4 assignment formatting.

Yeah you can just add a `// prettier-ignore` before it. The example in the docs is a matrix, and it's pretty much the only example I've ever run into myself.
https://prettier.io/docs/en/ignore.html

@hpinkos 

> Don't trailing commas kill old versions of IE?

You have to go all the way back to IE8 for this to be an issue. If we ever did need to build a version of Cesium for an ancient version of IE, we just change the setting locally and run prettier and all the trailing commas are gone.

> I guess I can still type a single quote all the time and prettier will just fix it for me on commit.

Yeah you can type it however you want and hit Alt-Shift-F (or configure VS Code to format on save) and it'll be reformatted. This is true of all code, not just quotes, and is a big reason prettier lets you not think about formatting at all. Copy, paste, don't bother reformatting for the new location, just hit the format hotkey. If it doesn't format you probably missed a closing brace (check the prettier output window).

**Comment by mramato** [comment]
> You might want to leave it but change the indent to 2. That way VS Code will know that new files should be indented with two spaces.

I actually already did this, just forgot to mention it.

**Comment by kring** [comment]
> Once you get used to how prettier indents things, the indentation tells you the structure / precedence of the code, 100% reliably every time.

To explain what I mean by the above, Prettier only ever does one of two things:
1. Puts all the elements on one line.
2. Puts each element on its own line.

It'll never mix these within a single statement and, say, put three elements on each of two lines. So with code like this:

```js
command.modelMatrix[13] -= 
  CesiumMath.sign(command.modelMatrix[13]) * 
  2.0 * 
  CesiumMath.PI * 
  projection.ellipsoid.maximumRadius; 
```

You can see at a glance (once you get used to it) that this is an arithmetic expression with 4 terms.

If that call to `CesiumMath.sign` had more arguments so it wouldn't fit in 80 characters, its parameters would be separated into one per line and all would be indented, so you could still see at a glance the expression with the four terms, once of which is a more complicated, indented sub-expression.

So I'm just doing the usual "justify the style I'm used to" thing here, of course, but my point is that the "weird" does have some reason behind it, it's not _just_ weird.

**Comment by mramato** [comment]
I will point out that @IanLilleyT first "bad" example would be fine if we inserted some line breaks before and after the `command.modelMatrix[13]` assignment.  Then it would be readable as 3 separate blocks.

### 80 column width
```js
if (defined(command) && model._mode === SceneMode.SCENE2D) {
  Matrix4.clone(nodeMatrix, command.modelMatrix);

  command.modelMatrix[13] -=
    CesiumMath.sign(command.modelMatrix[13]) *
    2.0 *
    CesiumMath.PI *
    projection.ellipsoid.maximumRadius;

  BoundingSphere.transform(
    primitiveCommand.boundingSphere,
    command.modelMatrix,
    command.boundingVolume
  );
}
```

vs

### 120 column width
```js
if (defined(command) && model._mode === SceneMode.SCENE2D) {
      Matrix4.clone(nodeMatrix, command.modelMatrix);
      command.modelMatrix[13] -=
        CesiumMath.sign(command.modelMatrix[13]) * 2.0 * CesiumMath.PI * projection.ellipsoid.maximumRadius;
      BoundingSphere.transform(primitiveCommand.boundingSphere, command.modelMatrix, command.boundingVolume);
    }
}
```

Shorter column width ultimately translates to more new lines for better break-up of the code.  It's just a different way of thinking. Prettier does not insert new lines so there may be a period of "clean-up" where we add some line breaks in the code or break out variables to make the code more readable.  

My official vote would be to go with the default 80 column width and curate from there. If we _really_ still hate it after a couple of months, we can always set it to something else and just run prettier. There's no risk associated with starting with the default.

Again, I totally get the push back here and there was a time in my life I would have done the same.  It will be a period of adjustment for everyone, including myself, but I think it's time to embrace change and give it a try.

**Comment by mramato** [comment]
Finally, I'll add that this is an incremental step. I would like to see introducing template literals, arrow functions, async/await and other niceties into Cesium's code base sometime this year. Once that happens, Cesium's verbosity would be greatly diminished and there would be even less indentation all around.

**Comment by aboutqx** [comment]
@mramato Be happy to see this changes. I've find https://github.com/babel/gulp-babel to be used with `gulp-uglify` in minifying es6 features. But i'm not familiar with rollup and workers, so i can't easily change the `gulpfile`. Could anyone add this so we can use template literals, arrow functions etc?

**Comment by mramato** [comment]
> Could anyone add this so we can use template literals, arrow functions etc?

@aboutqx While that's part of the long term plan, it is not related to this PR and will require additional work to make happen. Cesium needs to be very performance sensitive to do what it does, so just switching to babel may not be the best answer.  That doesn't stop you from using these features in your own code, just not as part of Cesium internals itself yet.

**Comment by hpinkos** [comment]
Alrighty, it looks like keeping the 80 column width and the rest of the defaults is the way to go.  This is definitely going to take some getting used to, but I think it's the right way to go so we won't waste any more time commenting about formatting in PRs.

@mramato did you check the size of the built zipfile and do a sanity check for IE?  I think those were the last two things I was concerned about.

**Comment by mramato** [comment]
1. Confirmed that everything works on IE 11.
2. Confirmed that combined file size is basically unchanged.

Spoke with @IanLilleyT offline and he's willing to give the defaults a try. (Thanks Ian!).

So I think we have consensus.  I'm going to merge in master and undo the CI check so that this branch passes. After that, I believe the steps are:

1. Merge this PR.
2. Immediately run the formatter and submit the changes to master.

If people have long-running branches that they need to update:

1. Merge in the merge commit for this PR (which will be one commit before the formatting changes)
2. Run format yourself on their branch
3. Merge in master and accept all of your changes (instead of worrying about resolving conflicts which will be format only)

The above means zero format-related merge issues. @kring can you please confirm these are the steps you've used in the past when switching to prettier.

**Comment by kring** [comment]
Not quite. If you do that exactly as you've written, you'll lose changes when you accept yours and it clobbers changes that happened in master after prettier came in. Here's the whole process:

First, we need two labels: `pre-prettier` and `post-prettier`. `pre-prettier` is the last master commit just before prettier is run, and `post-prettier` must be a commit that runs prettier and _nothing else_. It is vitally important that no other changes happen in `post-prettier` or you will be sad.

Then, when you need to merge post-prettiergeddon master into another branch, follow this process:

* Create a new branch for your merge: `git checkout -b whatever-merge`
* Merge the last commit before prettier into your branch: `git merge pre-prettier`. Resolve any conflicts as usual.
* Commit the merge above.
* Merge the commit that ran prettier on all the source files into your branch. This is likely to cause heaps of conflicts, but because you've already merged the last commit before prettier, and because the prettier commit only changes formatting, you can safely accept your version any time there is a conflict. Long story short, all you have to do is run: `git merge post-prettier --strategy=ours --no-commit`
* Run prettier on the result of the merge before committing it: `npm run prettier`
* Commit the merged result.
* Merge master into your branch in order to pick up any changes in master that happened after the prettiergeddon: `git merge origin/master`. Resolve any conflicts as normal; there shouldn't be any prettier-related conflicts.
* Commit and push your branch.
* Open a pull request of your merge branch into the original one. It should merge cleanly.

(adapted slightly from the version I wrote for Terria, which is here: https://docs.terria.io/guide/getting-started/#prettier)

**Comment by mramato** [comment]
Thanks @kring!  Once we merge this (probably tomorrow or Friday) I'm happy to help integrate anyone elses's branches to make this transition as pain-free as possible.

File: .travis.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ script:
   - npm --silent run deploy-status -- --status pending --message 'Waiting for build'
 
   - npm --silent run eslint
-  - npm --silent run prettier-check
+#  - npm --silent run prettier-check
```
Comment (kring) [inline .travis.yml:12]: Remember, don't uncomment this in the commit where you run prettier. It needs to be a separate commit _after_ the prettier run commit.

Comment (mramato) [reply .travis.yml:12]: Yep, on it.  Thanks for the reminder.

**Comment by aboutqx** [comment]
@mramato While es6 has been implemented on modern browsers, it seems that there may be not a performance loss but a speed up. 
As for IE11, it's time for it to disappear, some performance loss should be accepted for it.
These new features will make cesium code more easier to read and change, especially template literal for  the js + glsl code in files like `processPbrMaterials.js `.While now it would be a pain to change this kind of files even for cesium stuffs i think.
Once these feeatures be added, cesium would have more popularity in community.

**Comment by mramato** [comment]
> * Create a new branch for your merge: `git checkout -b whatever-merge`

@kring I'm not sure I understand why you recommend doing these steps in a separate branch.  It seems like an extra step for no reason.

**Comment by kring** [comment]
True, it's not really necessary.

**Comment by mramato** [comment]
Okay, cool.  I just pushed what are hopefully the final set of change.

@kring this should be ready, are you around to do a quick final look and merge?

**Comment by mramato** [comment]
Master has been formatted with prettier! Here are the step by step instructions for anyone that needs to update a branch.  I'm also on hand to help you with any branches if needed.

I tested these on my long-running `native-promises` branch which has a ton of code changes and it was absolutely painless.

1. Check out your branch: `git checkout <your_branch>`
2. Fetch latest everything: `git pull`
3. Update to change just before prettier: `git merge 05380a6d7f7869483f63d6a3bbbd89d68ebd1531`
4. Resolve any conflicts and commit your changes as usual.
5. Merge in prettier format changes only: `git merge 2fd0e8f7e4212bd1e7084299187f70597a6bbfd8 --strategy=ours --no-commit` (No files will appear checked out but you will be in "merge" mode.)
6. Format all code with prettier: `npm run prettier`
7. Commit these changes (which is technically a merge commit)
8. Get the rest of your changes from master as usual: `git merge origin/master`
9. Push all changes to your remote branch.

File: .editorconfig [hunk 3]
```diff
@@ -3,13 +3,7 @@ root = true
...
 [*]
 indent_style = space
-indent_size = 4
+indent_size = 2
 charset = utf-8
 trim_trailing_whitespace = true
 insert_final_newline = true
-
-[*.md]
-trim_trailing_whitespace = false
```
Comment (emackey) [inline .editorconfig:12]: Just FYI, trailing whitespace is significant in Markdown files, it's used for line breaks.

Comment (mramato) [reply .editorconfig:12]: Thanks @emackey, I did not know that.  Prettier actually does the right thing here, I put it back so that editors that don't know about MD specifically don't conflict with our formatting.

782b6065f0ec43e7840c7877d1203efdf9ced89e in master
